ledger/trie.go

```diff
@@ -207,44 +210,153 @@ func ToPath(pathBytes []byte) (Path, error) {
+func (k encKey) String() string {
+	return hex.EncodeToString(k)
+}
+
+// Equals compares this encoded key to another encoded key.
+// A nil encoded key is equivalent to an empty encoded key.
+func (k encKey) Equals(other encKey) bool {
+	return bytes.Equal(k, other)
+}
+
+// DeepCopy returns a deep copy of the encoded key.
+func (k encKey) DeepCopy() encKey {
+	newK := make([]byte, len(k))
+	copy(newK, k)
+	return newK
+}
+
 // Payload is the smallest immutable storable unit in ledger
 type Payload struct {
+	// encKey is key encoded using PayloadVersion.
+	// Version and type data are not encoded to save 3 bytes.
+	// NOTE: encKey translates to Key{} when encKey is
+	//       one of these three values:
+	//       nil, []byte{}, []byte{0,0}.
+	encKey encKey
+	value  Value
+}
+
+// serializablePayload is used to serialize ledger.Payload.
+// Encoder only serializes exported fields and ledger.Payload's
+// key and value fields are not exported.  So it is necessary to
+// use serializablePayload for encoding.
+type serializablePayload struct {
 	Key   Key
 	Value Value
 }
 
+// MarshalJSON returns JSON encoding of p.
+func (p Payload) MarshalJSON() ([]byte, error) {
+	k, err := p.Key()
+	if err != nil {
+		return nil, err
+	}
+	sp := serializablePayload{Key: k, Value: p.value}
+	return json.Marshal(sp)
+}
+
+// UnmarshalJSON unmarshals a JSON value of payload.
+func (p *Payload) UnmarshalJSON(b []byte) error {
+	if p == nil {
+		return errors.New("UnmarshalJSON on nil Payload")
+	}
+	var sp serializablePayload
+	if err := json.Unmarshal(b, &sp); err != nil {
+		return err
+	}
+	p.encKey = encodeKey(&sp.Key, PayloadVersion)
+	p.value = sp.Value
+	return nil
+}
+
+// MarshalCBOR returns CBOR encoding of p.
+func (p Payload) MarshalCBOR() ([]byte, error) {
+	k, err := p.Key()
+	if err != nil {
+		return nil, err
+	}
+	sp := serializablePayload{Key: k, Value: p.value}
+	return cbor.Marshal(sp)
+}
+
+// UnmarshalCBOR unmarshals a CBOR value of payload.
+func (p *Payload) UnmarshalCBOR(b []byte) error {
+	if p == nil {
+		return errors.New("UnmarshalCBOR on nil payload")
+	}
+	var sp serializablePayload
+	if err := cbor.Unmarshal(b, &sp); err != nil {
+		return err
+	}
+	p.encKey = encodeKey(&sp.Key, PayloadVersion)
+	p.value = sp.Value
+	return nil
+}
+
+// Key returns payload key.
+// Error indicates that ledger.Key can't be created from payload key, so
+// migration and reporting (known callers) should abort.
+// CAUTION: do not modify returned key because it shares underlying data with payload key.
+func (p *Payload) Key() (Key, error) {
+	if p == nil || len(p.encKey) == 0 {
+		return Key{}, nil
+	}
+	k, err := decodeKey(p.encKey, true, PayloadVersion)
+	if err != nil {
+		return Key{}, err
+	}
+	return *k, nil
+}
+
+// Value returns payload value.
+// CAUTION: do not modify returned value because it shares underlying data with payload value.
+func (p *Payload) Value() Value {
+	if p == nil {
+		return Value{}
+	}
+	return p.value
+}
+
```

Comment:
this is great, this can be a good start for the future PRs to remove deep copies.